### ISSUE TITLE
Regarding the issue of the low attack success rate of UGBA on GAT？

### ISSUE THREAD

**Issue by ChouBei28222** [ventr1c/UGBA#7]
"Hello! I am currently running UGBA on the Cora dataset using GAT as the target model, but I've noticed that the attack success rate (ASR) is extremely low. I have configured my environment exactly according to the method you provided. Could you please help me understand why this might be happening?"

Epoch 0, loss_inner: 1.87491, loss_target: 1.84807, homo loss: 0.49852 
acc_train_clean: 0.4177, ASR_train_attach: 0.0000, ASR_train_outter: 0.1533
Epoch 10, loss_inner: 0.93259, loss_target: 0.68400, homo loss: 0.10632 
acc_train_clean: 0.8780, ASR_train_attach: 1.0000, ASR_train_outter: 0.9981
Epoch 20, loss_inner: 0.48484, loss_target: 0.55034, homo loss: 0.11625 
acc_train_clean: 0.9335, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 30, loss_inner: 0.38923, loss_target: 0.61309, homo loss: 0.12993 
acc_train_clean: 0.9501, ASR_train_attach: 1.0000, ASR_train_outter: 0.9080
Epoch 40, loss_inner: 0.56011, loss_target: 3.74977, homo loss: 0.13420 
acc_train_clean: 0.9076, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 50, loss_inner: 0.99631, loss_target: 13.23223, homo loss: 0.14172 
acc_train_clean: 0.9002, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 60, loss_inner: 0.60226, loss_target: 1.93920, homo loss: 0.14919 
acc_train_clean: 0.9390, ASR_train_attach: 0.1000, ASR_train_outter: 1.0000
Epoch 70, loss_inner: 0.27402, loss_target: 0.19847, homo loss: 0.15475 
acc_train_clean: 0.9686, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 80, loss_inner: 0.26046, loss_target: 0.19174, homo loss: 0.15501 
acc_train_clean: 0.9704, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 90, loss_inner: 0.24778, loss_target: 0.19456, homo loss: 0.15343 
acc_train_clean: 0.9704, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 100, loss_inner: 0.23977, loss_target: 0.18054, homo loss: 0.15101 
acc_train_clean: 0.9760, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 110, loss_inner: 0.23116, loss_target: 0.17221, homo loss: 0.14730 
acc_train_clean: 0.9778, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 120, loss_inner: 0.21845, loss_target: 0.83062, homo loss: 0.14188 
acc_train_clean: 0.9778, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 130, loss_inner: 0.21485, loss_target: 0.85572, homo loss: 0.13134 
acc_train_clean: 0.9797, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 140, loss_inner: 0.21271, loss_target: 0.71106, homo loss: 0.11433 
acc_train_clean: 0.9778, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 150, loss_inner: 0.21026, loss_target: 0.86370, homo loss: 0.10842 
acc_train_clean: 0.9815, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 160, loss_inner: 0.20936, loss_target: 1.04892, homo loss: 0.10655 
acc_train_clean: 0.9834, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 170, loss_inner: 0.20553, loss_target: 0.98281, homo loss: 0.09465 
acc_train_clean: 0.9815, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 180, loss_inner: 0.23724, loss_target: 0.17700, homo loss: 0.10336 
acc_train_clean: 0.9649, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 190, loss_inner: 0.22372, loss_target: 1.64691, homo loss: 0.11446 
acc_train_clean: 0.9778, ASR_train_attach: 0.9000, ASR_train_outter: 0.9981
Epoch 200, loss_inner: 0.20530, loss_target: 1.13912, homo loss: 0.10265 
acc_train_clean: 0.9834, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 210, loss_inner: 0.20074, loss_target: 1.15127, homo loss: 0.09556 
acc_train_clean: 0.9852, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 220, loss_inner: 0.20057, loss_target: 1.10424, homo loss: 0.09370 
acc_train_clean: 0.9871, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 230, loss_inner: 0.19690, loss_target: 1.10946, homo loss: 0.09390 
acc_train_clean: 0.9889, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 240, loss_inner: 0.20918, loss_target: 1.67318, homo loss: 0.10400 
acc_train_clean: 0.9815, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 250, loss_inner: 0.24503, loss_target: 0.18280, homo loss: 0.12299 
acc_train_clean: 0.9778, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 260, loss_inner: 0.21209, loss_target: 0.63518, homo loss: 0.10885 
acc_train_clean: 0.9815, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 270, loss_inner: 0.20395, loss_target: 0.91733, homo loss: 0.09687 
acc_train_clean: 0.9871, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 280, loss_inner: 0.20053, loss_target: 0.93485, homo loss: 0.09320 
acc_train_clean: 0.9871, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 290, loss_inner: 0.19777, loss_target: 1.10915, homo loss: 0.09137 
acc_train_clean: 0.9908, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 300, loss_inner: 0.19605, loss_target: 1.12095, homo loss: 0.09372 
acc_train_clean: 0.9889, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 310, loss_inner: 0.19456, loss_target: 1.08339, homo loss: 0.09339 
acc_train_clean: 0.9908, ASR_train_attach: 1.0000, ASR_train_outter: 1.0000
Epoch 320, loss_inner: 0.26967, loss_target: 1.34549, homo loss: 0.12083 
acc_train_clean: 0.9649, ASR_train_attach: 0.3000, ASR_train_outter: 0.0900
Epoch 330, loss_inner: 0.24204, loss_target: 1.47127, homo loss: 0.11639 
acc_train_clean: 0.9741, ASR_train_attach: 0.5000, ASR_train_outter: 0.0747
Epoch 340, loss_inner: 0.26265, loss_target: 1.59602, homo loss: 0.13736 
acc_train_clean: 0.9741, ASR_train_attach: 0.6000, ASR_train_outter: 0.0556
Epoch 350, loss_inner: 0.23756, loss_target: 1.64162, homo loss: 0.10860 
acc_train_clean: 0.9760, ASR_train_attach: 0.7000, ASR_train_outter: 0.0613
Epoch 360, loss_inner: 0.23411, loss_target: 1.93235, homo loss: 0.10969 
acc_train_clean: 0.9778, ASR_train_attach: 0.7000, ASR_train_outter: 0.0536
Epoch 370, loss_inner: 0.22595, loss_target: 1.75135, homo loss: 0.10975 
acc_train_clean: 0.9797, ASR_train_attach: 0.7000, ASR_train_outter: 0.0498
Epoch 380, loss_inner: 0.22141, loss_target: 1.75662, homo loss: 0.10952 
acc_train_clean: 0.9797, ASR_train_attach: 0.8000, ASR_train_outter: 0.0498
Epoch 390, loss_inner: 0.23693, loss_target: 1.84757, homo loss: 0.10821 
acc_train_clean: 0.9760, ASR_train_attach: 0.7000, ASR_train_outter: 0.0517
load best weight based on the loss outter
precent of left attach nodes: 1.000
Namespace(cuda=True, dataset='Cora', debug=True, defense_mode='none', device_id=0, dis_weight=1, dropout=0.5, epochs=200, evaluate_mode='1by1', hidden=32, homo_boost_thrd=0.5, homo_loss_weight=50.0, inner=5, lr=0.01, model='GCN', no_cuda=False, prune_thr=0.1, seed=265, selection_method='cluster_degree', target_class=0, target_loss_weight=1, test_model='GAT', thrd=0.5, train_lr=0.01, trigger_size=3, trojan_epochs=400, use_vs_number=True, vs_number=10, vs_ratio=0, weight_decay=0.0005)
/data/mr/UGBA-main/models/GAT.py:61: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  self.labels = torch.tensor(labels, dtype=torch.long)
target class rate on Vs: 0.4000
accuracy on clean test nodes: 0.8444
Overall ASR: 0.1697
Flip ASR: 0.0578/225 nodes
Namespace(cuda=True, dataset='Cora', debug=True, defense_mode='none', device_id=0, dis_weight=1, dropout=0.5, epochs=200, evaluate_mode='1by1', hidden=32, homo_boost_thrd=0.5, homo_loss_weight=50.0, inner=5, lr=0.01, model='GCN', no_cuda=False, prune_thr=0.1, seed=125, selection_method='cluster_degree', target_class=0, target_loss_weight=1, test_model='GAT', thrd=0.5, train_lr=0.01, trigger_size=3, trojan_epochs=400, use_vs_number=True, vs_number=10, vs_ratio=0, weight_decay=0.0005)
target class rate on Vs: 0.4000
accuracy on clean test nodes: 0.8407
Overall ASR: 0.1734
Flip ASR: 0.0578/225 nodes
Namespace(cuda=True, dataset='Cora', debug=True, defense_mode='none', device_id=0, dis_weight=1, dropout=0.5, epochs=200, evaluate_mode='1by1', hidden=32, homo_boost_thrd=0.5, homo_loss_weight=50.0, inner=5, lr=0.01, model='GCN', no_cuda=False, prune_thr=0.1, seed=996, selection_method='cluster_degree', target_class=0, target_loss_weight=1, test_model='GAT', thrd=0.5, train_lr=0.01, trigger_size=3, trojan_epochs=400, use_vs_number=True, vs_number=10, vs_ratio=0, weight_decay=0.0005)
target class rate on Vs: 0.4000
accuracy on clean test nodes: 0.8519
Overall ASR: 0.1771
Flip ASR: 0.0533/225 nodes
Namespace(cuda=True, dataset='Cora', debug=True, defense_mode='none', device_id=0, dis_weight=1, dropout=0.5, epochs=200, evaluate_mode='1by1', hidden=32, homo_boost_thrd=0.5, homo_loss_weight=50.0, inner=5, lr=0.01, model='GCN', no_cuda=False, prune_thr=0.1, seed=527, selection_method='cluster_degree', target_class=0, target_loss_weight=1, test_model='GAT', thrd=0.5, train_lr=0.01, trigger_size=3, trojan_epochs=400, use_vs_number=True, vs_number=10, vs_ratio=0, weight_decay=0.0005)
target class rate on Vs: 0.3000
accuracy on clean test nodes: 0.8296
Overall ASR: 0.1439
Flip ASR: 0.0400/225 nodes
Namespace(cuda=True, dataset='Cora', debug=True, defense_mode='none', device_id=0, dis_weight=1, dropout=0.5, epochs=200, evaluate_mode='1by1', hidden=32, homo_boost_thrd=0.5, homo_loss_weight=50.0, inner=5, lr=0.01, model='GCN', no_cuda=False, prune_thr=0.1, seed=320, selection_method='cluster_degree', target_class=0, target_loss_weight=1, test_model='GAT', thrd=0.5, train_lr=0.01, trigger_size=3, trojan_epochs=400, use_vs_number=True, vs_number=10, vs_ratio=0, weight_decay=0.0005)
target class rate on Vs: 0.5000
accuracy on clean test nodes: 0.8556
Overall ASR: 0.2030
Flip ASR: 0.0756/225 nodes
Overall ASR: 0.1734 (GAT model, Seed: 320)
Overall Clean Accuracy: 0.8444
Total Overall ASR: 0.1734 
Total Clean Accuracy: 0.8444
